Telemetry SPI

ignasi35 opened the pull request:
References #52

ignasi35 commented on akka-projection-core/src/main/java/akka/projection/internal/spi/ProjectionMetrics.java:
```diff
@@ -0,0 +1,49 @@
+                   Throwable cause,
+                   ClassicActorSystemProvider systemProvider);
+
+}
```
This first proposal is quite similar to the Lagom `Projections Spi` except for the fact that this is not bound to a type (in Lagom, all instrumented projections have an `Offset`). Because there's no type that's known beforehand, instrumentation produces an `Object` on the first call on the sequence and that instance must be propagated with the `SteamElement`. Because we're only measuring 2 instants of the stream only `onProcessStart` and `onProcessComplete` need to share that `Object` instance.

Then, `onFailure` must be invoked when the stream fails.

This implementation can be invoked like this:

```scala
// based on the akka-projection-slick branch

  private val metrics: ProjectionMetrics = ???

  // Some funtion that may extract the insertionTime from the StreamElement (or return null)
  val insertionTimeExtractor: StreamElement => Long = _ => null
      
  val (killSwitch, stopped) = {
    Source
      .futureSource(futSource)
      // convert to a SourceWithContext
      .asSourceWithContext(elt => 
         metrics.onProcessStart(projectionId, insertionTimeExtractor(elt), systemProvider))
      .mapAsync(1) { elt => processElement(elt) }
      .mapContext(ctx => metrics.onProcessComplete(projectionId, ctx))
      // convert back to a normal Source
      .asSource
      .watchTermination() { (_, f: Future[Done]) =>
        f.recoverWith {
          case t: Throwable =>
            metrics.onFailure(projectionId, t, systemProvider)
            f
        }
      }
      .toMat(Sink.ignore)(Keep.both)
      .run()
  }
```

*NOTE*: the above example can be simplified to invoke `onProcessStart` and `onProcessComplete` immediately before and after the  `processElement` invocation (inside the `mapAsync`) as demoed in the snippet below. I think other projection implementations may have more stages between the `Source` creation and the `processElement` invocation to, for example, parse the `ByteString` from the wire, deserialize the payload, apply event evelutions,... . The purpose of the example above was to use the SPI using `SourceWithContext`.

Next, is a simplified usage of the `onProcessStart`/`onProcessComplete` proposal for the Slick projection PR:

```scala

Source
  .futureSource(futSource)
  .mapAsync(1) { elt =>
    val metricsContext = metrics.onProcessStart(projectionId, insertionTimeExtractor(elt), systemProvider)
    processElement(elt).map { x =>
      metrics.onProcessComplete(projectionId, metricsContext)
      x
    }
  }
  .watchTermination() { (_, f: Future[Done]) =>
    f.recoverWith {
      ... // etc
```

seglo replied:
How does this work when failures occur?  I see you have an `onFailure` in your contract, but it's unclear when it would be called. Does it need a reference to the object context too?

ignasi35 replied:
🤦🏼‍♂️
I cut too much out of the snippet. I think something like the following could work:

```scala
.watchTermination() { (_, f: Future[Done]) =>
  f.recoverWith {
    case t: Throwable =>
      metrics.onFailure(projectionId, t, systemProvider)
      f
  }
}
```

seglo replied:
I see. Should it accept an object context too like `onProcessComplete`?

ignasi35 replied:
> I see. Should it accept an object context too like `onProcessComplete`?

That is currently not possible. Note that `watchTermination` is not available in `XyzWithContextOps` so, in order to use it we must go back to vanilla `Source`s (dropping the context), unless there's a way to keep it around.

But, is it possible to take the element (or the `(elt,ctx)` tuple in this case) that caused the stream to fail and pass it along with the failure cause?

I don't recall needing a context on the Lagom SPI for telemetry. As long as I passed the `projectionId` and the `ActorSystem` all the necessary information was available to measure the failure (I think).

seglo replied:
> But, is it possible to take the element (or the `(elt,ctx)` tuple in this case) that caused the stream to fail and pass it along with the failure cause?

Not without user intervention.

> I don't recall needing a context on the Lagom SPI for telemetry. As long as I passed the `projectionId` and the `ActorSystem` all the necessary information was available to measure the failure (I think).

Ok, I was just curious :)

ignasi35 commented on akka-projection-core/src/main/java/akka/projection/internal/spi/SingleShotProjectionMetrics.java:
```diff
@@ -0,0 +1,40 @@
+                   String projectionKey,
+                   Throwable cause,
+                   ClassicActorSystemProvider systemProvider);
+
```
This is a simplification of the above in which the responsibility to measure periods falls on the projection implementor. Then, there's a single invocation required in the happy path. Here's an example usage:

```scala
Source
  .futureSource(futSource)
  .mapAsync(1) { elt =>
    val before = now
    processElement(elt).map { elt =>
      metrics.onProcessComplete(projectionId, insertionTimeExtractor(elt), period(before), systemProvider)
    }
  }
```

while this looks appealing it takes away some information from telemetry. Since there's no invocation before `processElement` there's no chance for telemetry to, for example, start an OpenTracing span or any future information cinnamon provides. Instead, just two `Long`s are passed in.

renatocaval replied:
I prefer the first approach, not only because it's better for Cinnamon, but also more obvious for implementors.

patriknw replied:
I agree, we should have the separate methods onProcessStart, onProcessComplete and so on

seglo commented on akka-projection-core/src/main/java/akka/projection/internal/spi/ProjectionMetrics.java:
```diff
@@ -0,0 +1,53 @@
+     * invoke this method in a stream stage as close to the stream stage where you read from the wire to
+     * measure the parsing, deserializing and other steps of the processing prior to the event handling.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param creationTimestamp if the element traversing the stream contains the creation time, provide it. Set
```
A `ProjectionId` that combines name and key will be merged as part of the slick implementation soon #35.

ignasi35 replied:
Yeah, I'm counting on that. ;-)

patriknw commented on akka-projection-core/src/main/java/akka/projection/internal/spi/ProjectionMetrics.java:
```diff
@@ -0,0 +1,53 @@
+
+import akka.actor.ClassicActorSystemProvider;
+
+public interface ProjectionMetrics {
```
no reason to write this in Java as far as I can see. no difficult signatures here

ignasi35 commented:
I dropped the `SingleShotMetrics` and rebased on top of `master` to use `PorjectionId`.

I rewrote the SPI in scala and added a Noop implementation.

patriknw commented on akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala:
```diff
@@ -0,0 +1,65 @@
+   *
+   * @param projectionId      the projection id
+   * @param creationTimestamp if the element traversing the stream contains the creation time, provide it. Set
+   *                          to null if the information is not available.
```
can't be `null`, `0L` is used as the magic undefined

patriknw commented on akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala:
```diff
@@ -0,0 +1,65 @@
+   * @param context      the contextual object returned by `onProcessStart`
+   * @return a contextual object. The returned instance may not be the received one.
+   */
+  def onProcessComplete(projectionId: ProjectionId, context: AnyRef): AnyRef
```
what is the returned context supposed to be used for here?

patriknw replied:
why does some of the methods have systemProvider but not all?

ignasi35 replied:
Cinnamon uses the context to propagate all necessary information from `onProcessStart` to `onProcessComplete`. That context is an arbitrary `Object` where each instrument may attach extra fields. 

`onProcessStart` and `onProcessFailure` get an `ActorSystemProvider` because  those two methods are entry points and will likely need the `ActorSystem` or the `Config` in it to setup any instrument required. `onProcessComplete`, OTOH, will get all necessary information via the context (an instrument may even decide to add a reference to the `ActorSystem` on the context if they need it).

patriknw replied:
Understood, but my question was about the **returned** context from ** onProcessComplete**

Is onProcessFailure an entry point? Looks mor like a result of processing.

ignasi35 replied:
Sorry, @patriknw I'm having one of those days...

`onProcessComplete` should return `Unit`.

ignasi35 replied:
> Is onProcessFailure an entry point? Looks mor like a result of processing.

`onProcessFailure` is an entry point in that it is not invoked _after_ any other method invocation. `onProcessStart`/`onProcessComplete` share a context but `onProcessFailure` is independant here (it may even be invoked before `onProcessStart` was ever invoked.

patriknw replied:
Ok, so it’s streamFailure. Not necessarily due to that the event handler failed processing

ignasi35 replied:
> Ok, so it’s streamFailure. Not necessarily due to that the event handler failed processing

Yes, and again, that's an inherited approach from the Lagom Projection Telemetry SPI. I'm not sure what information could a specific `onProcessFailure` method provide but we could add it and let each instrument decide if they need the ifnromation.

ignasi35 replied:
If we added `onProcessFailure` we'd be able to report the failure with context information and let the projection developer decide what to do with the failed envelope. In Lagom, a failure in handling caused a stream failure always but maybe we want the extra degree of freedom (maybe a failure processing an envelope is no big deal in some use cases and there's no need to fail the stream and even the offset is stored anyway).

patriknw commented on akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala:
```diff
@@ -0,0 +1,65 @@
+   * @param cause          the cause of the failure
+   * @param systemProvider a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+   */
+  def onFailure(projectionId: ProjectionId, cause: Throwable, systemProvider: ClassicActorSystemProvider): Unit
```
no context?

patriknw commented on akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala:
```diff
@@ -0,0 +1,65 @@
+}
+
+object NoopProjectionMetrics {
+  val instance = new NoopProjectionMetrics()
```
could even be `object NoopProjectionMetrics extends ProjectionMetrics` instead of an `instance`?

ignasi35 commented:
Thanks @patriknw for the more detailed review. My comments above about context propagation and the need (or not) of an `ActorSystem` reflect my current (limited) understanding of Cinnammon. I've asked the telemetry team on Slack to have a peek at the PR in case I missed something else.

ignasi35 commented:
I just realized this proposal is not using the best approach but a copy of what Lagom Projections SPI used.

I think we can have a `ProjectionsMetricsProvider` with a `create(projectionId: ProjectionId):ProjectionsMetrics` method which will eagerly create the instruments and simplify the signatures on the methods in `ProjectionsMetrics`.

ygree commented on akka-projection-core/src/main/scala/akka/projection/internal/spi/ProjectionMetrics.scala:
```diff
@@ -0,0 +1,66 @@
+/**
+ * INTERNAL API
+ */
+@InternalStableApi private[akka] trait ProjectionMetrics {
```
Having some hooks for a projection worker life cycle could be useful as well. It could provide some additional metrics such as number of active projections / workers.

Also it would be useful to provide an ActorRef or something else that represents a projection worker. It could be useful for attaching some metadata like instruments and synchronize its creation deletion with the projection worker life cycle.

ignasi35 commented:
Close in favor of #230
